### PR TITLE
Implement new `MomentumQNGOptimizer` compatible with (q)jit

### PR DESCRIPTION
**Pull request by SimoneGasperini**
The goal here is to implement a new Jax-based version of the `MomentumQNGOptimizer` compatible with `jax.jit` and `qml.qjit`. This PR depends on the q(jit) compatible version of the standard `QNGOptimizer` (see #7452 for more context).

Note that this optimization algorithm requires to keep track of a non-trivial internal state: in the current Autograd-based implementation, this state is stored in the optimizer's `accumulation` attribute as a `pennylane.numpy` array (see [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/optimize/momentum_qng.py#L122)). However, in this Jax-based version, the state needs to be carried over at each optimization step as it happens for the native `optax` optimizers. This is required to avoid having a Jax tracer as a class attribute when using (q)jit compilation.

[sc-92558]

### CODE REVIEW

File: pennylane/optimize/momentum_qng_qjit.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+            state = opt.init(params)
+            for _ in range(100):
+                params, state = opt.step(circuit, params, state)
+
+            return params
+
+    >>> workflow()
+    Array([ 3.14159265, -1.57079633], dtype=float64)
+
+    Make sure you are using the ``lightning.qubit`` device along with ``qml.qjit`` with ``autograph`` enabled.
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
```

File: pennylane/optimize/momentum_qng_qjit.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+# limitations under the License.
+"""Quantum natural gradient optimizer with momentum for Jax/Catalyst interface"""
+
+from pennylane import math
+
+from .qng_qjit import QNGOptimizerQJIT
+
+
+class MomentumQNGOptimizerQJIT(QNGOptimizerQJIT):
+    r"""Optax-like and ``jax.jit``/``qml.qjit``-compatible implementation of the ``MomentumQNGOptimizer``,
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
    r"""Optax-like and ``jax.jit``/``qml.qjit``-compatible implementation of the :class:`~.MomentumQNGOptimizer`,
```

File: pennylane/optimize/momentum_qng_qjit.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+              would be to combine all parameters into a single objective function argument.
+
+            - ``MomentumQNGOptimizerQJIT`` does not work correctly if there is any classical processing in the QNode circuit
+              (e.g., ``2 * theta`` as a gate parameter).
+
+    **Example:**
+
+    Consider a hybrid workflow to optimize an objective function defined by a quantum circuit.
+    To make the optimization faster, the entire workflow can be just-in-time compiled using
+    the ``qml.qjit`` decorator:
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
    the :func:`~.qjit` decorator:
```

File: pennylane/optimize/momentum_qng_qjit.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        opt = qml.MomentumQNGOptimizerQJIT(stepsize=0.1, momentum=0.2)
+        step = jax.jit(partial(opt.step, circuit))
+
+        params = jnp.array([0.1, 0.2])
+        state = opt.init(params)
+        for _ in range(100):
+            params, state = step(params, state)
+
+    >>> params
+    Array([ 3.14159265, -1.57079633], dtype=float64)
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]: We can leave this example out :) 

```suggestion
```

File: pennylane/optimize/momentum_qng_qjit.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    Keyword Args:
+        stepsize=0.01 (float): the user-defined stepsize hyperparameter
+        momentum=0.9 (float): the user-defined momentum coefficient hyperparameter
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
        stepsize=0.01 (float): the stepsize hyperparameter
        momentum=0.9 (float): the momentum coefficient hyperparameter
```

**Review by isaacdevlugt — CHANGES_REQUESTED** [review]
Nice! We also need a changelog entry 😁

File: pennylane/optimize/momentum_qng_qjit.py (hunk 4 above)
Comment (SimoneGasperini) [reply pennylane/optimize/momentum_qng_qjit.py]: Ok sure!

File: pennylane/optimize/momentum_qng_qjit.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+        Please be aware of the following:
+
+            - As with ``MomentumQNGOptimizer``, ``MomentumQNGOptimizerQJIT`` supports a single QNode to encode the objective function.
+
+            - ``MomentumQNGOptimizerQJIT`` does not support any QNode with multiple arguments. A potential workaround
+              would be to combine all parameters into a single objective function argument.
+
+            - ``MomentumQNGOptimizerQJIT`` does not work correctly if there is any classical processing in the QNode circuit
+              (e.g., ``2 * theta`` as a gate parameter).
```
Comment (mudit2812) [inline pennylane/optimize/momentum_qng_qjit.py]: Are these two restrictions also present for the `MomentumQNGOptimizer` or just the JAX version?

File: pennylane/optimize/momentum_qng_qjit.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+                params, state = opt.step(circuit, params, state)
+
+            return params
+
+    >>> workflow()
+    Array([ 3.14159265, -1.57079633], dtype=float64)
+
+
+
+    Keyword Args:
```
Comment (mudit2812) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion

    Keyword Args:
```

File: pennylane/optimize/momentum_qng_qjit.py (hunk 6 above)
Comment (SimoneGasperini) [reply pennylane/optimize/momentum_qng_qjit.py]: Just for the Jax version. This is actually due to current limitations in the `metric_tensor` computation

File: doc/releases/changelog-dev.md [hunk 8]
```diff
@@ -4,6 +4,11 @@
 
 <h3>New features since last release</h3>
 
+* Quantum just-in-time compilation to optimize parameterized hybrid workflows: the :class:`~.MomentumQNGOptimizerQJIT`
+  has been added to offer a `jax.jit`- and `qml.qjit`-compatible analogue to the existing :class:`~.MomentumQNGOptimizer`
+  with an Optax-like interface.
+  [(#7606)](https://github.com/PennyLaneAI/pennylane/pull/7606)
+
```
Comment (isaacdevlugt) [inline doc/releases/changelog-dev.md]:
````suggestion
* Leveraging quantum just-in-time compilation to optimize parameterized hybrid workflows with the quantum 
  momentum quantum natural gradient optimizer is now possible with the new 
  :class:`~.MomentumQNGOptimizerQJIT` optimizer. 
  [(#7606)](https://github.com/PennyLaneAI/pennylane/pull/7606)

  Similar to the :class:`~.QNGOptimizerQJIT` optimizer, :class:`~.MomentumQNGOptimizerQJIT` offers a 
  `jax.jit`- and `qml.qjit`-compatible analogue to the existing :class:`~.MomentumQNGOptimizer` with an 
  Optax-like interface:

  ```python
  import pennylane as qml
  import jax.numpy as jnp
  @qml.qjit(autograph=True)
  def workflow():
      dev = qml.device("lightning.qubit", wires=2)
      @qml.qnode(dev)
      def circuit(params):
          qml.RX(params[0], wires=0)
          qml.RY(params[1], wires=1)
          return qml.expval(qml.Z(0) + qml.X(1))
      
      opt = qml.MomentumQNGOptimizerQJIT(stepsize=0.1, momentum=0.2)
      params = jnp.array([0.1, 0.2])
      state = opt.init(params)
      
      for _ in range(100):
          params, state = opt.step(circuit, params, state)
      
      return params
  ```

  ```pycon
  >>> workflow()
  Array([ 3.14159265, -1.57079633], dtype=float64)
  ```
````

File: pennylane/optimize/momentum_qng.py [hunk 9]
```diff
@@ -81,6 +81,8 @@ class MomentumQNGOptimizer(QNGOptimizer):
         - `QAOA <https://github.com/borbysh/Momentum-QNG/blob/main/QAOA_depth4.ipynb>`__
         - `VQE <https://github.com/borbysh/Momentum-QNG/blob/main/portfolio_optimization.ipynb>`__
 
+        See :class:`~.MomentumQNGOptimizerQJIT` for an Optax-like and ``jax.jit``/``qml.qjit``-compatible implementation.
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng.py:84]: 👍

File: pennylane/optimize/momentum_qng_qjit.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+        Please be aware of the following:
+
+            - As with ``MomentumQNGOptimizer``, ``MomentumQNGOptimizerQJIT`` supports a single QNode to encode the objective function.
+
+            - ``MomentumQNGOptimizerQJIT`` does not support any QNode with multiple arguments. A potential workaround
+              would be to combine all parameters into a single objective function argument.
+
+            - ``MomentumQNGOptimizerQJIT`` does not work correctly if there is any classical processing in the QNode circuit
+              (e.g., ``2 * theta`` as a gate parameter).
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
        Please be aware of the following:

        - As with ``MomentumQNGOptimizer``, ``MomentumQNGOptimizerQJIT`` supports a single QNode to encode the objective function.

        - ``MomentumQNGOptimizerQJIT`` does not support any QNode with multiple arguments. A potential workaround
          would be to combine all parameters into a single objective function argument.

        - ``MomentumQNGOptimizerQJIT`` does not work correctly if there is any classical processing in the QNode circuit
          (e.g., ``2 * theta`` as a gate parameter).
```

File: pennylane/optimize/momentum_qng_qjit.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            - If ``None``, the full metric tensor is computed
+
+            - If ``"block-diag"``, the block-diagonal approximation is computed, reducing
+              the number of evaluated circuits significantly
+
+            - If ``"diag"``, the diagonal approximation is computed, slightly
+              reducing the classical overhead but not the quantum resources
+              (compared to ``"block-diag"``)
+
+        lam=0 (float): metric tensor regularization to be applied at each optimization step
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]: Nice! Apologies I missed this with the `QNGOptimizerQJIT` PR, but can we move the example below the code example? We'd need to fix the QNGOptimizer docstring as well :). Feel free to do that in a follow-up PR or in this one, doesn't matter to me.

**Review by isaacdevlugt — CHANGES_REQUESTED** [review]

File: pennylane/optimize/momentum_qng_qjit.py (hunk 11 above)
Comment (SimoneGasperini) [reply pennylane/optimize/momentum_qng_qjit.py]: Hey @isaacdevlugt, thanks for the comment! If I understand correctly what you mean, this should be done both for `QNGOptimizerQJIT` and `MomentumQNGOptimizerQJIT` docstrings in [8d6684e](https://github.com/PennyLaneAI/pennylane/pull/7606/commits/8d6684ecaa0b17ff624da388972ac7b2ef2157e4).

Comment (isaacdevlugt) [reply pennylane/optimize/momentum_qng_qjit.py]: Yes, both the `QNGOptimizerQJIT` and `MomentumQNGOptimizerQJIT` docstrings should have the call signature information (args / returns) before code examples. 

I'm just realizing that this is the case for many of the optimizers 🤔... Even the regular `QNGOptimizer`!

@SimoneGasperini I'll take care of shuffling things around elsewhere, but if you could switch the order here (in this PR) for `MomentumQNGOptimizerQJIT` that would be great 🚀. I'll open a separate PR with other docstring changes

File: pennylane/optimize/momentum_qng_qjit.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]: ah okay nice you already did it! 😅

**Review by isaacdevlugt — APPROVED** [review]

File: pennylane/optimize/momentum_qng_qjit.py (hunk 11 above)
Comment (SimoneGasperini) [reply pennylane/optimize/momentum_qng_qjit.py]: Ok cool, thanks 😃

File: doc/releases/changelog-dev.md [hunk 13]
```diff
@@ -4,6 +4,43 @@
...
+  Similar to the :class:`~.QNGOptimizerQJIT` optimizer, :class:`~.MomentumQNGOptimizerQJIT` offers a 
+  `jax.jit`- and `qml.qjit`-compatible analogue to the existing :class:`~.MomentumQNGOptimizer` with an 
+  Optax-like interface:
+
+  ```python
+  import pennylane as qml
+  import jax.numpy as jnp
+
+  @qml.qjit(autograph=True)
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md]: Once again a bit confused by the choice to `qjit` the whole workflow. I feel like it'd be better to showcase being able to qjit individual steps. If the user forgets the `autograph=True` part, this will have *substantial* performance penalties.  It seems like we are recommending something that they are likely to mess up. I'd prefer to recommend the more robust syntax.

The entire point of the current design is that you can qjit individual steps.

Comment (SimoneGasperini) [reply doc/releases/changelog-dev.md]: Mmh yes, I see (cc @isaacdevlugt)

Comment (albi3ro) [reply doc/releases/changelog-dev.md]: Same comment on the class docstring too.  I'd recommend showcasing jitting/ qjitting individual steps instead.

**Review by albi3ro — APPROVED** [review]
Looks great 🚀 Thanks for putting this together.

Docs comments non-blocking, but I would still recommend the examples using the recommended, robust syntax of jitting an individual step instead of the workflow as a whole.

**Comment by SimoneGasperini** [comment]
Thank you @albi3ro! 😃 

> Docs comments non-blocking, but I would still recommend the examples using the recommended, robust syntax of jitting an individual step instead of the workflow as a whole.

Waiting for @isaacdevlugt's final word for that :)

File: doc/releases/changelog-dev.md (hunk 13 above)
Comment (isaacdevlugt) [reply doc/releases/changelog-dev.md]: I think there are two things here: (1) qjit'ing the whole workflow and (2) ensuring autograph is present. 

WIth (1), this is how we expect users will interact with Catalyst for the most part—slap qjit on the whole thing 👏—it just works and makes everything faster 🚀. 

With (2), the goal is for autograph to be default in Catalyst, although it's not the default yet. I think it's better to show users a flag and marking it as `True` instead of having PL native control flow syntax in places. 

My preference is to leave the code example as is 🙂. But, I'd be happy to have a small note/disclaimer somewhere that says "make sure you use autograph here because ...". What do you think?

Comment (SimoneGasperini) [reply doc/releases/changelog-dev.md]: A disclaimer for `autograph` is already there for the `QNGOptimizerQJIT` (see [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/optimize/qng_qjit.py#L80)). We could add the same kind of note in the `MomentumQNGOptimizerQJIT` docstring as well.

Comment (SimoneGasperini) [reply doc/releases/changelog-dev.md]: I briefly discussed about this also with @Alex-Preciado and the recommended strategy here would be:

1. After the code example showing how to use `qjit` on the whole workflow, add to the docstring also an example to show its usage on a single optimization step (as we did in the `QNGOptimizerQJIT` docstring)
2. For both `QNGOptimizerQJIT` and `MomentumQNGOptimizerQJIT`, add a more exhaustive disclaimer/note pointing out the necessity of enabling `autograph` and a short explanation of why this is the case

Comment (albi3ro) [reply doc/releases/changelog-dev.md]: Just because it's how they will doesn't mean it's how they should. And in our examples we can provide what they should to influence how they will.

Comment (isaacdevlugt) [reply doc/releases/changelog-dev.md]: @SimoneGasperini yours and Alex's suggestion seems like a good plan 👍

Comment (isaacdevlugt) [reply doc/releases/changelog-dev.md]: @SimoneGasperini if we can, let's keep any examples/disclaimers very brief. While the concerns with users making a mistake are valid, the docstring for the optimizer should just be saying _what_ it does, not _how_ to use it, as that can start to venture off into other concepts that don't necessarily pertain to the optimizer's information 👍 

IMO, the "what it does" is best explained with the current code example, but a disclaimer can't hurt :)

Comment (SimoneGasperini) [reply doc/releases/changelog-dev.md]: Ok thank you @isaacdevlugt 👌🏽

I've just updated the disclaimer both for QNG and Momentum-QNG. Please take a look at [45e6f3d](https://github.com/PennyLaneAI/pennylane/pull/7606/commits/45e6f3d916c9910d70a24bf2c3b48c674f75db93) and feel free to change it as you like.

**Comment by SimoneGasperini** [comment]
@albi3ro @isaacdevlugt Code example updated!

File: pennylane/optimize/momentum_qng_qjit.py [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+            - If ``"diag"``, the diagonal approximation is computed, slightly
+              reducing the classical overhead but not the quantum resources
+              (compared to ``"block-diag"``)
+
+        lam (float): metric tensor regularization to be applied at each optimization step (default value: 0).
+
+    **Example:**
+
+    Consider a hybrid workflow to optimize an objective function defined by a quantum circuit.
+    To make the entire workflow faster, the update step function and the whole optimization function
```
Comment (isaacdevlugt) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
    To make the entire workflow faster, the update step and the whole optimization
```

**Review by isaacdevlugt — APPROVED** [review]

File: pennylane/optimize/momentum_qng_qjit.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
...
+        @qml.qjit
+        def update_step_qjit(i, args):
+            params, state = args
+            return opt.step(circuit, params, state)
+
+        @qml.qjit
+        def optimization_qjit(params, iters):
+            state = opt.init(params)
+            args = (params, state)
+            params, state = qml.for_loop(0, iters)(update_step_qjit)(args)
```
Comment (albi3ro) [inline pennylane/optimize/momentum_qng_qjit.py]:
```suggestion
            params, state = qml.for_loop(iters)(update_step_qjit)(args)
```
